client/checker_client/cli.py

```diff
@@ -0,0 +1,229 @@
+import click
+
+from checker_client import checker as checker_lib
+import pytezos
+from pathlib import Path
+from dataclasses import dataclass
+from marshmallow import Schema, fields
+from marshmallow.decorators import post_load
+import portpicker
+import json
+
+CONFIG_FILE = Path.home().joinpath(Path(".checker"))
+
+
+@dataclass
+class Config:
+    path: str
+    sandbox_port: int = 18731
+    sandbox_container: str = "checker-sandbox-node"
+    tezos_address: str = "http://127.0.0.1"
+    tezos_port: int = 18731
+    tezos_key: str = "edsk3RFfvaFaxbHx8BMtEW1rKQcPtDML3LXjNqMNLCzC3wLC1bWbAt"  # bob's private key from "edobox info"
+    ctez_address: str = ""
+    oracle_address: str = ""
+    checker_address: str = ""
+
+    @staticmethod
+    def load(file: Path):
+        with file.open("r") as f:
+            return ConfigSchema().loads(f.read())
+
+    def dump(self):
+        with open(self.path, "w") as f:
+            f.write(ConfigSchema().dumps(self))
+
+
+class ConfigSchema(Schema):
+    path = fields.String(required=True)
+    sandbox_container = fields.String()
+    sandbox_port = fields.Int()
+    sandbox_container = fields.String()
+    tezos_address = fields.String()
+    tezos_port = fields.Int()
+    tezos_key = fields.String()
+    ctez_address = fields.String()
+    oracle_address = fields.String()
+    checker_address = fields.String()
+
+    @post_load
+    def make(self, data, **kwargs):
+        return Config(**data)
+
+
+@click.group()
+@click.pass_context
+@click.option(
+    "--config",
+    "config_file",
+    default=CONFIG_FILE,
+    help="path to CLI config file",
+    show_default=True,
+)
+def cli(ctx, config_file):
+    """Checker command line utilities
+
+    By default this tool caches options for things like addresses in the
+    specified config file to make chaining calls easier.
+    """
+    config_file = Path(config_file)
+    if config_file.exists():
+        config = Config.load(config_file)
+    else:
+        config = Config(path=config_file)
+        config.dump()
+    ctx.obj = config
+
+
+@cli.group()
+def sandbox():
+    """Interact with the local Tezos sandbox for Checker"""
+    pass
+
+
+@sandbox.command()
+@click.pass_obj
+def start(config: Config):
+    """Starts the sandbox"""
+    if checker_lib.is_sandbox_container_running(Config.sandbox_container):
+        click.echo(f"Container {Config.sandbox_container} is already running.")
+    else:
+        click.echo("Starting sandbox container...")
+        if not portpicker.is_port_free(config.sandbox_port):
+            click.echo(f"Selected host port {config.sandbox_port} for sandbox.")
+            config.sandbox_port = portpicker.pick_unused_port()
```

Comment:
I don't really understand this part. If the sandbox_port is not free, do we implicitly select a new one? This is a bit weird because if the user explicitly has made a choice we should respect that. Also, why are we printing the old port, then picking a new one?

Actually, since this command line only spins up one sandbox, the `portpicker` might be unnecessary and we can hardcode a port number.

Reply:
Good catch on the misplaced echo command, and that makes sense. Originally you actually couldn't explicitly specify the sandbox port, so I figured it would be easiest to let it pick one automatically. I agree that it makes sense to let people manually specify one if there is a conflict (since there is only one container at the moment anyways) and updated things accordingly.